File: sdks/node/src/health.ts
// Standalone liveness/readiness probes over a Queue handle. Usable on their own —
// wire them into any HTTP framework, a CLI check, or a container probe; the
// dashboard and the REST helpers are just callers.

import type { Queue } from "./queue";

/** Liveness payload: the process answered, nothing else is asserted. */
export interface HealthReport {
  status: "ok";
}

/** Worker check: how many workers heartbeated recently. */
export interface WorkersCheck {
  count: number;
  status: "ok" | "none";
}

/** Worker-resource check: names of resources not reporting `healthy`. */
export interface ResourcesCheck {
  count: number;
  unhealthy: string[];
  status: "ok" | "degraded";
}

/**
 * Per-dependency results. A check that threw is reported as the string
 * `"error: <message>"` instead of its object, and forces `degraded`.
 * `resources` is absent when no worker advertises any resource.
 */
export interface ReadinessChecks {
  storage: string;
  workers: WorkersCheck | string;
  resources?: ResourcesCheck | string;
}

/** Readiness payload: `ready` only when every check passed. */
export interface ReadinessReport {
  status: "ready" | "degraded";
  checks: ReadinessChecks;
}

/** Basic liveness check — always ok. */
export function checkHealth(): HealthReport {
  return { status: "ok" };
}

/**
 * Readiness: storage reachable, workers alive, resources healthy. Never
 * throws — a failing dependency lands in `checks` and degrades the status,
 * so a probe endpoint can always answer.
 */
export async function checkReadiness(queue: Queue): Promise<ReadinessReport> {
  const checks: ReadinessChecks = { storage: "ok", workers: { count: 0, status: "none" } };
  let allOk = true;

  try {
    await queue.stats();
  } catch (error) {
    checks.storage = `error: ${String(error)}`;
    allOk = false;
  }

  try {
    const workers = await queue.listWorkers();
    checks.workers = { count: workers.length, status: workers.length > 0 ? "ok" : "none" };
  } catch (error) {
    checks.workers = `error: ${String(error)}`;
    allOk = false;
  }

  try {
    const resources = await resourceStatus(queue);
    if (resources.length > 0) {
      const unhealthy = resources.filter((r) => r.health !== "healthy").map((r) => r.name);
      checks.resources = {
        count: resources.length,
        unhealthy,
        status: unhealthy.length > 0 ? "degraded" : "ok",
      };
      if (unhealthy.length > 0) {
        allOk = false;
      }
    }
  } catch (error) {
    checks.resources = `error: ${String(error)}`;
    allOk = false;
  }

  return { status: allOk ? "ready" : "degraded", checks };
}

/** One worker resource as seen across every live worker's heartbeat. */
export interface ResourceStatusEntry {
  name: string;
  scope: string;
  health: string;
  init_duration_ms: number;
  recreations: number;
  depends_on: string[];
}

/**
 * Worker-resource health aggregated from heartbeat snapshots: any `unhealthy`
 * report wins; all `healthy` → `healthy`; anything else (a `degraded` report,
 * or an unrecognised one) → `degraded`; advertised but not yet reported →
 * `not_initialized`.
 */
export async function resourceStatus(queue: Queue): Promise<ResourceStatusEntry[]> {
  const observed = new Map<string, string[]>();
  const advertised = new Set<string>();
  for (const worker of await queue.listWorkers()) {
    for (const name of parseJsonArray(worker.resources)) {
      advertised.add(name);
    }
    const report = parseJsonObject(worker.resourceHealth);
    for (const [name, healthValue] of Object.entries(report)) {
      const entries = observed.get(name) ?? [];
      entries.push(String(healthValue).toLowerCase());
      observed.set(name, entries);
    }
  }

  const names = new Set([...advertised, ...observed.keys()]);
  return [...names].sort().map((name) => {
    const reports = observed.get(name) ?? [];
    // The worst report wins, so one sick worker is never averaged away and a
    // fleet that is merely degraded is never escalated to unhealthy.
    let healthState = "not_initialized";
    if (reports.some((r) => r === "unhealthy")) {
      healthState = "unhealthy";
    } else if (reports.length > 0) {
      healthState = reports.every((r) => r === "healthy") ? "healthy" : "degraded";
    }
    return {
      name,
      scope: "worker",
      health: healthState,
      init_duration_ms: 0,
      recreations: 0,
      depends_on: [],
    };
  });
}

function parseJsonArray(raw: string | undefined | null): string[] {
  if (!raw) {
    return [];
  }
  try {
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? parsed.map(String) : [];
  } catch {
    return [];
  }
}

function parseJsonObject(raw: string | undefined | null): Record<string, unknown> {
  if (!raw) {
    return {};
  }
  try {
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : {};
  } catch {
    return {};
  }
}

File: sdks/java/src/main/java/org/byteveda/flexiq/health/Health.java
package org.byteveda.flexiq.health;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.function.BinaryOperator;
import org.byteveda.flexiq.FlexiQ;
import org.byteveda.flexiq.model.WorkerInfo;
import org.jspecify.annotations.Nullable;

/** Liveness and readiness probes that can be wired into any HTTP framework or container probe. */
public final class Health {
    private static final ObjectMapper JSON = new ObjectMapper();

    private Health() {}

    /**
     * Basic liveness check — always ok.
     *
     * @return an {@code ok} report; a process that could not answer would never reach here
     */
    public static HealthReport check() {
        return new HealthReport("ok");
    }

    /**
     * Readiness: storage reachable, workers alive, resources healthy. Never
     * throws — a failing dependency lands in its own check and degrades the
     * status, so a probe endpoint can always answer.
     *
     * @param queue the queue to probe
     * @return the report, {@code degraded} rather than thrown when a dependency fails
     */
    public static ReadinessReport readiness(FlexiQ queue) {
        String storage = "ok";
        boolean allOk = true;
        try {
            queue.stats();
        } catch (RuntimeException e) {
            storage = "error: " + e;
            allOk = false;
        }

        WorkersCheck workers;
        try {
            int count = queue.listWorkers().size();
            workers = new WorkersCheck(count, count > 0 ? "ok" : "none", null);
        } catch (RuntimeException e) {
            workers = new WorkersCheck(0, "error", "error: " + e);
            allOk = false;
        }

        ResourcesCheck resources = null;
        try {
            List<ResourceStatusEntry> entries = resourceStatus(queue);
            // Absent, not empty, when nothing advertises a resource — there is
            // nothing to be ready for.
            if (!entries.isEmpty()) {
                List<String> unhealthy = entries.stream()
                        .filter(entry -> !"healthy".equals(entry.health()))
                        .map(ResourceStatusEntry::name)
                        .toList();
                resources =
                        new ResourcesCheck(entries.size(), unhealthy, unhealthy.isEmpty() ? "ok" : "degraded", null);
                allOk = allOk && unhealthy.isEmpty();
            }
        } catch (RuntimeException e) {
            resources = new ResourcesCheck(0, List.of(), "error", "error: " + e);
            allOk = false;
        }

        return new ReadinessReport(allOk ? "ready" : "degraded", storage, workers, resources);
    }

    /**
     * Per-resource status aggregated across workers, derived from each worker's
     * advertised resources plus the health its heartbeat reported. A resource a
     * worker advertises but never reports on is {@code not_initialized}; when
     * workers disagree, the worst health wins.
     *
     * @param queue the queue whose workers are polled
     * @return one entry per resource any live worker advertises or reports on, by name
     */
    public static List<ResourceStatusEntry> resourceStatus(FlexiQ queue) {
        Map<String, Integer> reported = new TreeMap<>();
        Set<String> advertised = new TreeSet<>();
        for (WorkerInfo worker : queue.listWorkers()) {
            advertised.addAll(parseStringList(worker.resources));
            // BinaryOperator.maxBy keeps the merge on boxed Integers — Math::max
            // would funnel both arguments through an unchecked Integer->int unboxing.
            parseMap(worker.resourceHealth)
                    .forEach((name, value) -> reported.merge(
                            name, severity(String.valueOf(value)), BinaryOperator.maxBy(Comparator.naturalOrder())));
        }

        Set<String> names = new TreeSet<>(advertised);
        names.addAll(reported.keySet());
        List<ResourceStatusEntry> out = new ArrayList<>(names.size());
        for (String name : names) {
            Integer worst = reported.get(name);
            String health = worst == null ? "not_initialized" : healthName(worst);
            out.add(new ResourceStatusEntry(name, "worker", health, 0, 0, List.of()));
        }
        return out;
    }

    private static int severity(String health) {
        return switch (health) {
            case "unhealthy" -> 2;
            case "degraded" -> 1;
            default -> 0;
        };
    }

    private static String healthName(int severity) {
        return switch (severity) {
            case 2 -> "unhealthy";
            case 1 -> "degraded";
            default -> "healthy";
        };
    }

    /** A heartbeat's JSON payloads are advisory: unreadable ones read as absent. */
    private static List<String> parseStringList(@Nullable String json) {
        if (json == null || json.isEmpty()) {
            return List.of();
        }
        try {
            return JSON.readValue(json, new TypeReference<List<String>>() {});
        } catch (Exception e) {
            return List.of();
        }
    }

    private static Map<String, Object> parseMap(@Nullable String json) {
        if (json == null || json.isEmpty()) {
            return Map.of();
        }
        try {
            return JSON.readValue(json, new TypeReference<Map<String, Object>>() {});
        } catch (Exception e) {
            return Map.of();
        }
    }
}
